pandas/core/arrays/categorical.py

```diff
@@ -2521,42 +2289,43 @@ def _replace(self, *, to_replace, value, inplace: bool = False):
         cat = self if inplace else self.copy()
 
-        # build a dict of (to replace -> value) pairs
-        if is_list_like(to_replace):
-            # if to_replace is list-like and value is scalar
-            replace_dict = {replace_value: value for replace_value in to_replace}
-        else:
-            # if both to_replace and value are scalar
-            replace_dict = {to_replace: value}
-
         # other cases, like if both to_replace and value are list-like or if
         # to_replace is a dict, are handled separately in NDFrame
-        for replace_value, new_value in replace_dict.items():
-            if new_value == replace_value:
+        if not is_list_like(to_replace):
+            to_replace = [to_replace]
+
+        categories = cat.categories.tolist()
+        removals = set()
+        for replace_value in to_replace:
+            if value == replace_value:
+                continue
+            if replace_value not in cat.categories:
                 continue
-            if replace_value in cat.categories:
-                if isna(new_value):
-                    with catch_warnings():
-                        simplefilter("ignore")
-                        cat.remove_categories(replace_value, inplace=True)
-                    continue
-
-                categories = cat.categories.tolist()
-                index = categories.index(replace_value)
-
-                if new_value in cat.categories:
-                    value_index = categories.index(new_value)
-                    cat._codes[cat._codes == index] = value_index
-                    with catch_warnings():
-                        simplefilter("ignore")
-                        cat.remove_categories(replace_value, inplace=True)
-                else:
-                    categories[index] = new_value
-                    with catch_warnings():
-                        simplefilter("ignore")
-                        cat.rename_categories(categories, inplace=True)
+            if isna(value):
+                removals.add(replace_value)
+                continue
+
+            index = categories.index(replace_value)
+
+            if value in cat.categories:
+                value_index = categories.index(value)
+                cat._codes[cat._codes == index] = value_index
+                removals.add(replace_value)
+            else:
+                categories[index] = value
+                cat._set_categories(categories)
+
+        if len(removals):
+            new_categories = [c for c in categories if c not in removals]
+            new_dtype = CategoricalDtype(new_categories, ordered=self.dtype.ordered)
+            codes = recode_for_categories(
+                cat.codes, cat.categories, new_dtype.categories
+            )
+            NDArrayBacked.__init__(cat, codes, new_dtype)
+
         if not inplace:
             return cat
+        NDArrayBacked.__init__(self, cat._ndarray, cat.dtype)
```

Comment:
why is this line after the `return cat`?

Reply:
Thanks for catching. That line is no longer needed. It was needed when the code was creating new objects to ensure `self` was updated when `inplace=True`. Now that the operations above are all inplace there is no need for it. I've removed it.